geode-core/src/main/java/org/apache/geode/internal/cache/AbstractRegionEntry.java

```diff
@@ -1327,16 +1327,21 @@ public Object prepareValueForCache(RegionEntryContext r,
         if (isSerialized) {
           if (event != null && event.getCachedSerializedNewValue() != null) {
             data = event.getCachedSerializedNewValue();
-          } else if (val instanceof CachedDeserializable) {
-            data = ((CachedDeserializable) val).getSerializedValue();
-          } else if (val instanceof PdxInstance) {
-            try {
-              data = ((ConvertableToBytes) val).toBytes();
-            } catch (IOException e) {
-              throw new PdxSerializationException("Could not convert " + val + " to bytes", e);
-            }
           } else {
-            data = EntryEventImpl.serialize(val);
+            if (val instanceof CachedDeserializable) {
+              data = ((CachedDeserializable) val).getSerializedValue();
+            } else if (val instanceof PdxInstance) {
+              try {
+                data = ((ConvertableToBytes) val).toBytes();
+              } catch (IOException e) {
+                throw new PdxSerializationException("Could not convert " + val + " to bytes", e);
+              }
+            } else {
+              data = EntryEventImpl.serialize(val);
+            }
+          }
+          if (event != null) {
+            event.setCachedSerializedNewValue(data);
```

Comment:
Move this block (i.e. setCachedSerializedNewValue) up one line. It should be in the new "else" block. Currently it is in the "isSerialized" block.